### ISSUE TITLE
TensorForest: Performance issues in time and space

### ISSUE THREAD

**Issue by whatbeg** [tensorflow/tensorflow#24386]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux CentOS 7
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary, pip install tensorflow==1.12.0
- TensorFlow version (use command below): 1.12.0
- Python version: 3.6.4
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: No usage
- GPU model and memory: No usage
- Machine RAM and #Cores: **64GB RAM and 12 CPU Cores** 

**Describe the current behavior**

Currently, I trained TensorForest on a relatively large train data (1.5M examples, 1024 features) with 500 trees, max_nodes. Moreover, I compared its performance with Scikit-learn implementation.
It seems that the TensorForest is inferior to Scikit-Learn ExtraTrees implementation both in time and space.

**In terms of space costs**, training a TensorForest forest  consists of 500 trees with 10k nodes will throw the following exception in my machine (64GB RAM and 12 CPU Cores).
```
ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[1500000,500,10] and type float on /job:localhost/replica:0/task:0/device:CPU:0 by allocator cpu
	 [[node stack (defined at /home/experiment/huqiu/anaconda3/lib/python3.6/site-packages/tensorflow/contrib/tensor_forest/python/tensor_forest.py:519)  = Pack[N=500, T=DT_FLOAT, axis=1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](TreePredictionsV4, TreePredictionsV4_1, TreePredictionsV4_2, TreePredictionsV4_3, TreePredictionsV4_4, ...... TreePredictionsV4_499)
```
But Scikit-learn based extra trees classifier only costs up to **18GB** of memory.

This issue makes it unfeasible to train TensorForest on large datasets or large number of trees.
As the paper said, It seems that TensorForest should not have such a big performance gap compared with scikit-learn, right? 
Thus I wonder whether it is my mis-use?

**In terms of time costs**,  training a TensorForest forest  consists of **only 2** trees with 10k nodes will cost **1235 second** with 116 training steps on the machine. But Scikit-learn based implementation costs only **20** seconds.
This 60-fold gap is clearly unreasonable, right?

**Describe the expected behavior**

1. Isn't it an online algorithm? It seems that the memory consumption of the TensorForest is relatively large?
2. The original paper said that, in large dataset (like HIGGS, 11M examples, 28 features), TensorForest with 100 trees and 10k nodes per tree trains in about **one percent** of the time taken by scikit-learn, even without taking advantage of distributed training. But I cannot reproduce that fast training speed.

Thus I am sincerely looking forward to your help or explanations.
Great, Great, Great thanks : )
@yupbank @yongtang @tensorflower-gardener

**Code to reproduce the issue**

Complete Code Snippet, you can run it directely without any modification.
============================== Space Cost Comparison =========================
TensorForest:
```
import time
import tensorflow as tf
tf.logging.set_verbosity(tf.logging.INFO)
tf.keras.backend.set_image_data_format('channels_first')

(x_train, y_train), _ = tf.keras.datasets.cifar10.load_data()
x_train = x_train.reshape((-1, 32*32)).astype('float32')   # produce more examples
y_train = y_train.repeat(3, axis=0).astype('int')
x_train = x_train.repeat(10, axis=0)     # repeat 10 times to produce more examples
y_train = y_train.repeat(10, axis=0).reshape(-1)

print(x_train.shape, y_train.shape)      # totally 1500000 examples
# (1500000, 1024) (1500000,)

start_time = time.time()
est_args = {'num_classes': 10, 'num_features': 1024, 'regression': False,
                   'num_trees': 500, 'max_nodes': 10000,
                   'base_random_seed': 0}

params = tf.contrib.tensor_forest.python.tensor_forest.ForestHParams(**est_args)
estimator = tf.contrib.tensor_forest.client.random_forest.TensorForestEstimator(params)
estimator.fit(x_train, y_train)
print("Time Cost: {}".format(time.time() - start_time))

# Exception: ResourceExhaustedError
```

Scikit-learn:
```
import time
import tensorflow as tf
tf.logging.set_verbosity(tf.logging.INFO)
tf.keras.backend.set_image_data_format('channels_first')

(x_train, y_train), _ = tf.keras.datasets.cifar10.load_data()
x_train = x_train.reshape((-1, 32*32)).astype('float32')   # produce more examples
y_train = y_train.repeat(3, axis=0).astype('int')
x_train = x_train.repeat(10, axis=0)     # repeat 10 times to produce more examples
y_train = y_train.repeat(10, axis=0).reshape(-1)

print(x_train.shape, y_train.shape)      # totally 1500000 examples
# (1500000, 1024) (1500000,)

start_time = time.time()
est_args = {'num_classes': 10, 'num_features': 1024, 'regression': False,
                   'num_trees': 500, 'max_nodes': 10000,
                   'base_random_seed': 0}

from sklearn.ensemble import ExtraTreesClassifier
estimator = ExtraTreesClassifier(n_estimators=500, n_jobs=-1)
estimator.fit(x_train, y_train)
print("Time Cost: {}".format(time.time() - start_time))

# Time Cost: 885.44 (seconds)
```

============================== Time Cost Comparison =========================
TensorForest:
```
import time
import tensorflow as tf
tf.logging.set_verbosity(tf.logging.INFO)
tf.keras.backend.set_image_data_format('channels_first')

(x_train, y_train), _ = tf.keras.datasets.cifar10.load_data()
x_train = x_train.reshape((-1, 32*32)).astype('float32')   # produce more examples
y_train = y_train.repeat(3, axis=0).astype('int')
x_train = x_train.repeat(10, axis=0)     # repeat 10 times to produce more examples
y_train = y_train.repeat(10, axis=0).reshape(-1)

print(x_train.shape, y_train.shape)      # totally 1500000 examples
# (1500000, 1024) (1500000,)

start_time = time.time()
est_args = {'num_classes': 10, 'num_features': 1024, 'regression': False,
                   'num_trees': 2, 'max_nodes': 10000,
                   'base_random_seed': 0}

params = tf.contrib.tensor_forest.python.tensor_forest.ForestHParams(**est_args)
estimator = tf.contrib.tensor_forest.client.random_forest.TensorForestEstimator(params)
estimator.fit(x_train, y_train)
print("Time Cost: {}".format(time.time() - start_time))

# Time Cost: 1235 (seconds)
```

Scikit-learn:
```
import time
import tensorflow as tf
tf.logging.set_verbosity(tf.logging.INFO)
tf.keras.backend.set_image_data_format('channels_first')

(x_train, y_train), _ = tf.keras.datasets.cifar10.load_data()
x_train = x_train.reshape((-1, 32*32)).astype('float32')   # produce more examples
y_train = y_train.repeat(3, axis=0).astype('int')
x_train = x_train.repeat(10, axis=0)     # repeat 10 times to produce more examples
y_train = y_train.repeat(10, axis=0).reshape(-1)

print(x_train.shape, y_train.shape)      # totally 1500000 examples
# (1500000, 1024) (1500000,)

start_time = time.time()
est_args = {'num_classes': 10, 'num_features': 1024, 'regression': False,
                   'num_trees': 500, 'max_nodes': 10000,
                   'base_random_seed': 0}

from sklearn.ensemble import ExtraTreesClassifier
estimator = ExtraTreesClassifier(n_estimators=500, n_jobs=-1)
estimator.fit(x_train, y_train)
print("Time Cost: {}".format(time.time() - start_time))

# Time Cost: 20.39 (seconds)
```

@yupbank

**Comment by yupbank** [comment]
Thanks for the issue I’ll take a look at it

**Comment by whatbeg** [comment]
@yupbank thanks : ) 
I think I know where the problem occurs.
The official guide said that using `estimator = tf.contrib.tensor_forest.client.random_forest.TensorForestEstimator(params)`
to train and predict.
But the arguments passing brought the problem.
If we do not provide the batch size when fit, the default batch size should be 128.
But if we directly use `estimator.fit(x_train, y_train)`, the batch size will be None and finally becomes the size of full dataset.
Thus, we actually do full batch training!

To fix it, I currently wrap the TensorForestEstimator in the SKCompat.
That is to say, I use:
```
from tensorflow.contrib.learn.python.learn.estimators.estimator import SKCompat
params = tf.contrib.tensor_forest.python.tensor_forest.ForestHParams(**est_args)
estimator = SKCompat(tf.contrib.tensor_forest.client.random_forest.TensorForestEstimator(params))
estimator.fit(x_train, y_train)
```
This code will use batchsize = 128 to train an online TensorForest.

Thank you : )

**Comment by yupbank** [comment]
Thanks for the explanation! Yes the batch size matters here. 

A lot of the API you are using is being deprecated, but we havnt merge the core estimator code. So yeah ...

**Comment by whatbeg** [comment]
@yupbank yes, I know some APIs are being deprecated, will they appear in the next version? in other forms? e.g. convert TensorForestEstimator to an estimator of `tf.estimators` ?

**Comment by yupbank** [comment]
Yes yes I’m working on that

**Comment by whatbeg** [comment]
Thanks.
Wishing for your significant work!
